libcontainer/setns_init_linux.go

```diff
@@ -49,6 +49,13 @@ func (l *linuxSetnsInit) Init() error {
 			}
 		}
 	}
+
+	// Clean the RLIMIT_NOFILE cache in go runtime.
+	// Issue: https://github.com/opencontainers/runc/issues/4195
+	if containsRlimit(l.config.Rlimits, unix.RLIMIT_NOFILE) {
+		system.ClearRlimitNofileCache()
+	}
+
```

Comment:
Can you explain why we need the `containsRlimit`? Would be easier to always remove the cache as we don't really need it in runc.

Reply:
Also, maybe we can move code to a single place, e.g. `system.Exec`.

```diff
diff --git a/libcontainer/system/linux.go b/libcontainer/system/linux.go
index 0b3b4da3..949a5b2a 100644
--- a/libcontainer/system/linux.go
+++ b/libcontainer/system/linux.go
@@ -9,6 +9,7 @@ import (
        "os"
        "os/exec"
        "strconv"
+       "sync/atomic"
        "syscall"
        "unsafe"
 
@@ -44,7 +45,18 @@ func Execv(cmd string, args []string, env []string) error {
        return Exec(name, args, env)
 }
 
+//go:linkname syscallOrigRlimitNofile syscall.origRlimitNofile
+var syscallOrigRlimitNofile atomic.Pointer[syscall.Rlimit]
+
 func Exec(cmd string, args []string, env []string) error {
+       // As reported in issue 4195, the new version of go runtime (since
+       // 1.19) caches rlimit-nofile on startup, and restores it in Exec.
+       // In runc, this causes the rlimit-nofile setting by the parent
+       // process for the container to become invalid. The solution is to
+       // clear this cache (for that, we need a go:linkname hack to access
+       // the private var `origRlimitNofile` in the syscall package).
+       syscallOrigRlimitNofile.Store(nil)
+
        for {
                err := unix.Exec(cmd, args, env)
                if err != unix.EINTR {
```

(i've also modified the comment a bit, for clarity).

Reply:
> Can you explain why we need the `containsRlimit`? Would be easier to always remove the cache as we don't really need it in runc.

Because if we run a container without a specific nofile rlimit, we should not clear the origRlimitNofile. Consider the system's [curr, max]=[1024, 65536], after we run `runc exec/run/create`, the [curr, max] has been changed to [65536, 65536] by go runtime, and the origRlimitNofile stores [1024, 65536], if we clear this cache, the nofile limit for the container will become wrong.

Reply:
> Also, maybe we can move code to a single place, e.g. `system.Exec`.

Consider the above reason, it is difficulty to move code to here.

Reply:
I think I wrote a comment earlier but it got lost.

We can put this call to `containerInit()`:
```diff
diff --git a/libcontainer/init_linux.go b/libcontainer/init_linux.go
index bd89b253..523259d4 100644
--- a/libcontainer/init_linux.go
+++ b/libcontainer/init_linux.go
@@ -223,6 +223,10 @@ func containerInit(t initType, config *initConfig, pipe *syncSocket, consoleSock
                return err
        }
 
+       if containsRlimit(config.Rlimits, unix.RLIMIT_NOFILE) {
+               system.ClearRlimitNofileCache()
+       }
+
        switch t {
        case initSetns:
                i := &linuxSetnsInit{
```

Reply:
Thanks.